File: Fundamentals_of_Software_Design/1_Basic_skills/Перетасовка/Простые числа/main.py
def sieve_of_eratosthenes(N):
    # Массив для отметки простых чисел
    is_prime = [True] * (N + 1)
    
    # 0 и 1 не являются простыми числами
    is_prime[0] = is_prime[1] = False
    
    # Алгоритм решета Эратосфена
    for i in range(2, int(N**0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, N + 1, i):
                is_prime[j] = False
    
    # Выводим простые числа
    primes = [i for i in range(2, N + 1) if is_prime[i]]
    return primes

# Ввод числа N
N = int(input("Введите число N: "))

# Получаем простые числа и выводим их
primes = sieve_of_eratosthenes(N)
print(f"Простые числа, меньшие или равные {N}:")
print(" ".join(map(str, primes)))

File: Fundamentals_of_Software_Design/1_Basic_skills/Перетасовка/Простые числа/main.c
#include <stdio.h>
#include <stdbool.h>

void sieve_of_eratosthenes(int N) {
    // Массив для отметки простых чисел
    bool is_prime[N + 1];
    
    // Изначально все числа считаются простыми
    for (int i = 0; i <= N; i++) {
        is_prime[i] = true;
    }
    
    // 0 и 1 не являются простыми числами
    is_prime[0] = is_prime[1] = false;
    
    // Алгоритм решета Эратосфена
    for (int i = 2; i * i <= N; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= N; j += i) {
                is_prime[j] = false;
            }
        }
    }
    
    // Выводим простые числа
    for (int i = 2; i <= N; i++) {
        if (is_prime[i]) {
            printf("%d ", i);
        }
    }
    printf("\n");
}

int main() {
    int N;
    printf("Введите число N: ");
    scanf("%d", &N);

    printf("Простые числа, меньшие или равные %d:\n", N);
    sieve_of_eratosthenes(N);

    return 0;
}
